CASSANDRA-19946 Uses ParameterizedClass for all auth-related implementations

tiagomlalves opened the pull request:
Uses ParameterizedClass for IAuthorizer, INetworkAuthorizer, and IRoleManager implementations enabling those to be configurable from parameters specified directly in cassandra.yaml.

Opportunistically makes consistent the initialization and setting of all auth-related implementations and removes code duplication.

Notes:
 * IInternodeAuthenticator implementations are expected to be set as default directly in DatabaseDescriptor instead of done via AuthConfig.applyAuth(). This is assumed in tests and client logic. For instance ReconnectableSnitchHelperTest fetches this authenticator before calling daemonInitialize(). Also, BulkLoader fetches this directly when creating an outbound connection.
 * Changing this behavior in BulkLoader causes a cascade of changes. First, one would need to add AuthConfig.applyAuth() directly in clientInitialize() and then would need to enable all the additional classes that this change causes. Long term this might be justified as if we're depending on the auth logic to be properly set. However, in the context of this change it was decided to postpone further changes.

Thanks for sending a pull request! Here are some tips if you're new here:
 
 * Ensure you have added or run the [appropriate tests](https://cassandra.apache.org/_/development/testing.html) for your PR.
 * Be sure to keep the PR description updated to reflect all changes.
 * Write your PR title to summarize what this PR proposes.
 * If possible, provide a concise example to reproduce the issue for a faster review.
 * Read our [contributor guidelines](https://cassandra.apache.org/_/development/index.html)
 * If you're making a documentation change, see our [guide to documentation contribution](https://cassandra.apache.org/_/development/documentation.html)
 
Commit messages should follow the following format:

```
<One sentence description, usually Jira title or CHANGES.txt summary>

<Optional lengthier description (context on patch)>

patch by <Authors>; reviewed by <Reviewers> for CASSANDRA-#####

Co-authored-by: Name1 <email1>
Co-authored-by: Name2 <email2>

```

The [Cassandra Jira](https://issues.apache.org/jira/projects/CASSANDRA/issues/)

tiagomlalves committed:
CASSANDRA-19946 Uses ParameterizedClass for all auth-related implementations

Uses ParameterizedClass for IAuthorizer, INetworkAuthorizer, and
IRoleManager implementations enabling those to be configurable from
parameters specified directly in cassandra.yaml.

Opportunistically makes consistent the initialization and setting of
all auth-related implementations and removes code duplication.

Notes:
 * IInternodeAuthenticator implementations are expected to be set as
   default directly in DatabaseDescriptor instead of done via
   AuthConfig.applyAuth(). This is assumed in tests and client logic.
   For instance ReconnectableSnitchHelperTest fetches this authenticator
   before calling daemonInitialize(). Also, BulkLoader fetches this
   directly when creating an outbound connection.
 * Changing this behavior in BulkLoader causes a cascade of changes.
   First, one would need to add AuthConfig.applyAuth() directly in
   clientInitialize() and then would need to enable all the additional
   classes that this change causes. Long term this might be justified
   as if we're depending on the auth logic to be properly set. However,
   in the context of this change it was decided to postpone further
   changes.

tiagomlalves committed:
CASSANDRA-19946 Adds test for 5.1 configuration

Adds 5.1 configuration test and fixes existent 5.0 configuration test
which was changed incorrectly (5.1 configuration test should have been
created when 5.0 was released).

Fix done in 5.0 configuration test:
 * Removed parameters `max_certificate_validity_period` and
   `certificate_validity_warn_threshold` introduced in CASSANDRA-1895
   for Cassandra 5.1 only.

Changes done in 5.1 configuration test (compared to 5.0):
 * Added above parameters removed from 5.0 configuration.
 * Removed `native_transport_port_ssl` which was removed from 5.0
   in CASSANDRA-18959.
 * Modified `require_client_auth` from Boolean to String as this
   was changed for 5.0 and doesn't need to be checked for 5.1 in
   CASSANDRA-18811

tiagomlalves commented:
Triggered CI `#49` which shows 5 test failures, none of them related with these changes.

michaelsembwever commented on conf/cassandra_latest.yaml:
```diff
@@ -180,37 +184,57 @@ authenticator:
 #   increase system_auth keyspace replication factor if you use this authorizer.
-authorizer: AllowAllAuthorizer
+authorizer:
+  class_name: org.apache.cassandra.auth.AllowAllAuthorizer
```
all these changes in cassandra_latest.yaml  can also be applied to cassandra.yaml (this being the trunk patch, and this not being anything new and experimental)

tiagomlalves replied:
done.

michaelsembwever replied:
oh, i'm late to notice…

all these class_name values should be in short form.
e.g. here `class_name: AllowAllAuthorizer`

it's a standard in our config that default classes can be in short form.
The PR's code already supports this, so it's just the yaml settings that need adjusting.  (both yaml in this PR, and the latest yaml in the 5.0 PR)

tiagomlalves replied:
No worries. Yes, code changes allow both FQCN and short form and unfortunately they were using a mix of both. I'm fine using either way, and since the FQCN seem to be the predominant one I've changed it that. Should we standardize on both `cassandra.yaml` and `cassandra_latest.yaml` for all 5.0 and trunk on the usage of the short form? If not, please clarify which forms should be used where.

michaelsembwever replied:
it's easy to be wrong about these things, but my understanding is the precedence is to use short names for actual values, and also FQNs in the comments so that it's intuitive that both can be used.

michaelsembwever replied:
to clarify, in this PR and the 5.0 PR, on lines where `class_name: ` is introduced then use short class names.

ignore lines that are not otherwise being touched in these PRs.

tiagomlalves replied:
Agreed to use short class names in all places and fixing other places not following that convention to prevent confusion in the future.

michaelsembwever reviewed (approved):
LGTM.  just the one comment about bringing changes from cassandra_latest.yaml also over into cassandra.yaml

tiagomlalves committed:
CASSANDRA-19946 Updates cassandra configuration files

Updates cassandra configuration files making it explicit the
possibility of specifying parameters for auth-related implementations.

tiagomlalves commented:
@michaelsembwever all changes done.

tiagomlalves committed:
CASSANDRA-19946 Standardize auth-related confs

Standardize auth-related configurations by using a short-form with
the class name instead of the full qualified class name containing
the package name.

michaelsembwever commented:
Committed with https://github.com/apache/cassandra/commit/1defa0851f7c1a557ebf76a566070308c222fa16
